### PR TITLE
WT-9036 Add WT_CONFIG_PRESERVE_QUOTE macro

### PR DESCRIPTION
**Pull request by ddanderson**
This is an alternative to #7982 - a low impact solution.

### CODE REVIEW

**Comment by sueloverso** [comment]
Of the three styles, this one wins my vote. It is the easiest to understand and keeps the changes well located and clear. I think it will be easiest for others to have correct code in the future. (I.e. someone writing some other function in the future might forget to use `config_next_quoted` and use `config_next` and reintroduce the bug.)

**Comment by ddanderson** [comment]
@sueloverso, I'm fine with this one - I like that it's small and easy to understand.

The argument for #7982 for me is kind of philosophical:
- parsing quoted strings is arguably broken, you don't get out what you put in
- we have a public interface to the parser
- therefore we should fix any bugs in a way that fixes it publicly.

But I think that's more of a philosophical than a practical argument, and we could argue the "arguable" part!  The fact is that really nobody, except some stray extension code here and there, is probably going to ever use our config parser, it's doubtful that anyone would notice this, and if they ever did, they _could_ work around it with macros.  So I'm happy to go with this one - it's the least work, there's no external impact.

@keitharnoldsmith - do you have thoughts on this?

**Comment by keitharnoldsmith** [comment]
I agree with @sueloverso that the simplicity and encapsulation of this approach makes it appealing.

I understand the philosophical argument for #7982, but I'm not a fan of adding complexity to support a use case we don't think anyone uses.  If this is a concern, I would suggest updating the documentation to point out this shortcoming and then if/when it becomes an issue for a WT user, we can address it.

**Comment by ddanderson** [comment]
Thanks, @keitharnoldsmith, we're all in agreement, this is the way to go.

@sueloverso and @AlexeyAnisimovAU, can you review this please?

File: src/config/config_collapse.c [hunk 1]
```diff
@@ -121,14 +117,10 @@ __config_merge_scan(
...
-        if (k.type == WT_CONFIG_ITEM_STRING) {
-            --k.str;
-            k.len += 2;
-        }
-        if (v.type == WT_CONFIG_ITEM_STRING) {
-            --v.str;
-            v.len += 2;
-        }
+        if (k.type == WT_CONFIG_ITEM_STRING)
+            WT_CONFIG_PRESERVE_QUOTES(&k);
```
Comment (AlexeyAnisimovAU) [inline src/config/config_collapse.c:121]: This looks a bit suspicious to me. What if we are handling the first key in the config string here? `WT_CONFIG_PRESERVE_QUOTES` will try to access `str[-1]` which is before the first symbol in the config string. There's no problem with the handling of values below and in `tiered_handle.c`. Do we really need to support quoted keys?

**Review by AlexeyAnisimovAU — COMMENTED** [review]
Looks nice and clean. But I have one concern.

File: src/config/config_collapse.c (hunk 1 above)
Comment (ddanderson) [reply src/config/config_collapse.c:121]: Nice catch!  I think we do have to handle this case - as a matter of fact the code it replaces expects every key and value to be quoted (it doesn't even check).  This code is not the case that is failing on the ticket (that's `tiered_handle.c`).  I'm going to code it in a safer way, swapping the order of checks.

**Comment by ddanderson** [comment]
@AlexeyAnisimovAU, can you please give this another review?

**Review by AlexeyAnisimovAU — APPROVED** [review]
This is much safer now. One suggestion from me, otherwise, LGTM!

File: src/include/config.h [hunk 2]
```diff
@@ -46,6 +46,21 @@ struct __wt_config_parser_impl {
 
 #define WT_CONFIG_ITEM_STATIC_INIT(n) static const WT_CONFIG_ITEM n = {"", 0, 0, WT_CONFIG_ITEM_NUM}
 
+/*
+ * If double quotes surround the string, then expand the string to include them. This is always
+ * called in the context of keys or values returned by the configuration parser. The character after
+ * the string must be at a valid memory address, and checking just that one is sufficient. If it is
+ * a double quote, then the character before must be as well, by the rules of the tokenizer.
+ */
+#define WT_CONFIG_PRESERVE_QUOTES(session, item)        \
```
[the remark this answers is not in the document]
Comment (ddanderson) [reply src/include/config.h:55]: Thanks, I've made that change too.

**Comment by ddanderson** [comment]
@sueloverso, can you take another look at this please?

**Review by sueloverso — APPROVED** [review]
LGTM

**Pull request merged**